task/control-plane-agent/src/mgs_sidecar.rs

```diff
@@ -90,13 +142,32 @@ impl MgsHandler {
         if self.common.sp_update.is_preparing() {
             Some(sys_get_timer().now + 1)
         } else {
-            None
+            // Find the soonest LockState::UnlockedUntil time (if present)
+            self.locked
+                .iter()
+                .flat_map(|k| match k {
+                    LockState::UnlockedUntil(t) => Some(*t),
+                    _ => None,
+                })
+                .min()
         }
     }
 
     pub(crate) fn handle_timer_fired(&mut self) {
         // This is a no-op if we're not preparing for an SP update.
         self.common.sp_update.step_preparation();
+
+        let now = sys_get_timer().now;
+        for (i, k) in self.locked.clone().iter().enumerate() {
+            if let LockState::UnlockedUntil(lock_at) = k {
+                if now >= *lock_at {
```

Comment:
fwiw, this has recreated most of multitimer. (You'd need an enum rather than an integer naming your vlans to use multitimer.)